executor/tests/test_progress.py

```diff
@@ -0,0 +1,288 @@
+                          progress_watcher.send_progress_update_helper, ce.send_message)
+
+        self.assertTrue(progress_complete_event.isSet())
+        self.assertEqual(4, progress_watcher.method_call_count())
+        self.assertEqual([0, 1, 2, 3], progress_watcher.get_progress_states())
+
+    def test_match_progress_update(self):
+
+        def match_progress_update(input_string):
+            progress_regex_string = '\^\^\^\^JOB-PROGRESS: (\d*)(?: )?(.*)'
+            return cp.ProgressWatcher.match_progress_update(progress_regex_string, input_string)
+
+        self.assertIsNone(match_progress_update("One percent complete"))
+        self.assertIsNone(match_progress_update("^^^^JOB-PROGRESS 1 One percent complete"))
+        self.assertIsNone(match_progress_update("JOB-PROGRESS: 1 One percent complete"))
+        self.assertEqual(('1', ''),
+                         match_progress_update("^^^^JOB-PROGRESS: 1"))
+        self.assertEqual(('1', 'One percent complete'),
+                         match_progress_update("^^^^JOB-PROGRESS: 1 One percent complete"))
+        self.assertEqual(('50', 'Fifty percent complete'),
+                         match_progress_update("^^^^JOB-PROGRESS: 50 Fifty percent complete"))
+        # Fractions in progress update are not supported
+        self.assertEqual(('2', '.0 Two percent complete'),
+                         match_progress_update("^^^^JOB-PROGRESS: 2.0 Two percent complete"))
+
+    def test_send_progress_update(self):
+        driver = FakeMesosExecutorDriver()
+        task_id = get_random_task_id()
+        max_message_length = 100
+        poll_interval_ms = 100
+
+        progress_updater = cp.ProgressUpdater(poll_interval_ms)
+        progress_data_0 = {'progress-message': 'Progress message-0'}
+        progress_updater.send_progress_update(driver, task_id, max_message_length, progress_data_0, ce.send_message)
+
+        self.assertEqual(1, driver.method_call_count())
+        actual_encoded_message_0 = driver.messages[0]
+        expected_message_0 = {'progress-message': 'Progress message-0', 'task-id': task_id}
+        assert_message(expected_message_0, actual_encoded_message_0)
+
+        progress_data_1 = {'progress-message': 'Progress message-1'}
+        progress_updater.send_progress_update(driver, task_id, max_message_length, progress_data_1, ce.send_message)
+
+        self.assertEqual(1, driver.method_call_count())
+
+        time.sleep(poll_interval_ms / 1000.0)
+        progress_data_2 = {'progress-message': 'Progress message-2'}
+        progress_updater.send_progress_update(driver, task_id, max_message_length, progress_data_2, ce.send_message)
+
+        self.assertEqual(2, driver.method_call_count())
+        actual_encoded_message_2 = driver.messages[1]
+        expected_message_2 = {'progress-message': 'Progress message-2', 'task-id': task_id}
+        assert_message(expected_message_2, actual_encoded_message_2)
+
+    def test_progress_watcher_tail(self):
```

Comment:
Can you add a test case where you send a LOT of lines? I think the yields will handle this gracefully but want to be sure. Also, if a LOT of lines were sent out, would the progress updater just be way behind?